File: ArxivExpress/ArxivExpress/Features/SearchArticles/Data/SearchQueryRepository.cs
using System;
using System.IO;
using System.Linq;
using System.Xml.Linq;

namespace ArxivExpress.Features.SearchArticles.Data
{
    public class SearchQueryRepository
    {
        private static SearchQueryRepository _instance;

        private SearchQueryRepository()
        {
        }

        public static SearchQueryRepository GetInstance()
        {
            if (_instance == null)
            {
                _instance = new SearchQueryRepository();
            }

            return _instance;
        }

        private string GetFileName()
        {
            return "search_query.xml";
        }

        private string GetSearchQueryElementName()
        {
            return "SearchQuery";
        }

        private string GetFilePath()
        {
            return Path.Combine(Environment.GetFolderPath(
                Environment.SpecialFolder.LocalApplicationData), GetFileName());
        }

        public SearchQuery LoadSearchQuery()
        {
            var filePath = GetFilePath();

            if (File.Exists(filePath))
            {
                var xml = XDocument.Load(filePath);
                var list = (
                    from searchQuery
                    in xml.Descendants(GetSearchQueryElementName())
                    select new SearchQuery
                    {
                        SearchTerm = searchQuery.Attribute("SearchTerm")?.Value,
                        Prefix = searchQuery.Attribute("Prefix")?.Value,
                        ResultsPerPage = searchQuery.Attribute("ResultsPerPage")?.Value,

                        SortByRelevance = bool.Parse(searchQuery.Attribute("SortByRelevance")?.Value),
                        SortByLastUpdatedDate = bool.Parse(searchQuery.Attribute("SortByLastUpdatedDate")?.Value),
                        SortBySubmittedDate = bool.Parse(searchQuery.Attribute("SortBySubmittedDate")?.Value),

                        SortOrderAscending = bool.Parse(searchQuery.Attribute("SortOrderAscending")?.Value),
                        SortOrderDescending = bool.Parse(searchQuery.Attribute("SortOrderDescending")?.Value)
                    }
                ).ToList();

                if (list.Count != 0)
                {
                    return list[0];
                }
            }

            return new SearchQuery();
        }

        public void SaveSearchQuery(SearchQuery searchQuery)
        {
            var filePath = GetFilePath();

            var xml = new XDocument();

            var attributes = new object[8];
            attributes[0] = new XAttribute("SearchTerm", searchQuery.SearchTerm);
            attributes[1] = new XAttribute("Prefix", searchQuery.Prefix);
            attributes[2] = new XAttribute("ResultsPerPage", searchQuery.ResultsPerPage);
            attributes[3] = new XAttribute("SortByRelevance", searchQuery.SortByRelevance);
            attributes[4] = new XAttribute("SortByLastUpdatedDate", searchQuery.SortByLastUpdatedDate);
            attributes[5] = new XAttribute("SortBySubmittedDate", searchQuery.SortBySubmittedDate);
            attributes[6] = new XAttribute("SortOrderAscending", searchQuery.SortOrderAscending);
            attributes[7] = new XAttribute("SortOrderDescending", searchQuery.SortOrderDescending);

            xml.Add(new XElement(GetSearchQueryElementName(), attributes));
            xml.Save(filePath);
        }
    }
}

File: ArxivExpressFlutter/lib/features/search_articles/data/search_query_repository.dart
// ****************************************************************************
//    File "search_query_repository.dart"
//    Copyright © Dmitry Morozov 2025
//    Flutter reimplementation of ArxivExpress Xamarin app
// ****************************************************************************

import 'package:shared_preferences/shared_preferences.dart';
import 'search_query.dart';

class SearchQueryRepository {
  static SearchQueryRepository? _instance;
  
  static SearchQueryRepository getInstance() {
    _instance ??= SearchQueryRepository._internal();
    return _instance!;
  }
  
  SearchQueryRepository._internal();

  Future<SearchQuery> loadSearchQuery() async {
    final prefs = await SharedPreferences.getInstance();
    
    final searchQuery = SearchQuery();
    searchQuery.searchTerm = prefs.getString('searchTerm') ?? '';
    searchQuery.prefix = prefs.getString('prefix') ?? 'all';
    searchQuery.resultsPerPage = prefs.getString('resultsPerPage') ?? '25';
    searchQuery.sortByRelevance = prefs.getBool('sortByRelevance') ?? false;
    searchQuery.sortByLastUpdatedDate = prefs.getBool('sortByLastUpdatedDate') ?? false;
    searchQuery.sortBySubmittedDate = prefs.getBool('sortBySubmittedDate') ?? true;
    searchQuery.sortOrderAscending = prefs.getBool('sortOrderAscending') ?? false;
    searchQuery.sortOrderDescending = prefs.getBool('sortOrderDescending') ?? true;
    searchQuery.pageNumber = prefs.getInt('pageNumber') ?? 0;
    
    return searchQuery;
  }

  Future<void> saveSearchQuery(SearchQuery searchQuery) async {
    final prefs = await SharedPreferences.getInstance();
    
    await prefs.setString('searchTerm', searchQuery.searchTerm);
    await prefs.setString('prefix', searchQuery.prefix);
    await prefs.setString('resultsPerPage', searchQuery.resultsPerPage);
    await prefs.setBool('sortByRelevance', searchQuery.sortByRelevance);
    await prefs.setBool('sortByLastUpdatedDate', searchQuery.sortByLastUpdatedDate);
    await prefs.setBool('sortBySubmittedDate', searchQuery.sortBySubmittedDate);
    await prefs.setBool('sortOrderAscending', searchQuery.sortOrderAscending);
    await prefs.setBool('sortOrderDescending', searchQuery.sortOrderDescending);
    await prefs.setInt('pageNumber', searchQuery.pageNumber);
  }
}